# Import Data

In [1]:
import pandas as pd

df = pd.read_csv('/content/part4_metrics.csv')
df.head()

design       workload  credits       ipc  ipc_per_credit       cpi  \
0  design_a      basicmath      820  0.190587        0.000232  5.246957   
1  design_a      bitcounts      820  1.046242        0.001276  0.955802   
2  design_a          qsort      820  0.526698        0.000642  1.898622   
3  design_a    susan_edges      820  1.043843        0.001273  0.957999   
4  design_a  susan_corners      820  1.093927        0.001334  0.914138   

   branchMispredicts  branch_mispredict_rate  commitSquashedInsts  \
0          2411853.0                1.501748           27972490.0   
1          2294635.0                3.918540           13078880.0   
2           483503.0               12.364862            8397359.0   
3           189157.0                3.526751            3044549.0   
4            74823.0                3.704214            1220810.0   

   commit_utilization_pct  ...  numIssuedDist_2  numIssuedDist_mean  \
0                33.51960  ...     1.272662e+09            0.677052   
1                91.94590  ...     4.904820e+08            1.850623   
2                50.47415  ...     3.700607e+07            1.099728   
3                78.58160  ...     3.607454e+07            1.591284   
4                80.95800  ...     1.401746e+07            1.636336   

   numIssuedDist_total      simInsts        simOps  simSeconds      simTicks  \
0         4.538739e+09  1.606031e+09  3.039729e+09    8.426774  8.426774e+12   
1         5.596423e+08  5.855841e+08  1.024916e+09    0.559702  5.597024e+11   
2         7.423938e+07  3.910298e+07  7.363242e+07    0.074242  7.424180e+10   
3         5.137404e+07  5.363491e+07  8.002585e+07    0.051382  5.138219e+10   
4         1.845757e+07  2.019943e+07  2.958084e+07    0.018465  1.846506e+10   

   speculation_overhead_pct  squashedInstsExamined  squashedInstsIssued  
0                  1.741716             39804461.0            1356018.0  
1                  2.233476             13079161.0             600707.0  
2                 21.474983              8928918.0             261072.0  
3                  5.676431              3043689.0             211450.0  
4                  6.043785              1224336.0             177808.0  

[5 rows x 59 columns]

# GRAPH 1: IPC Comparison Across Designs and Workloads

# Task
Generate a grouped bar chart using `plotly.express` to compare IPC across different designs and workloads, coloring the bars by `design` and setting `workload` on the x-axis and `ipc` on the y-axis, with `barmode='group'`. Define a custom color map, prepare legend labels to include credit costs for each design, and apply chart customizations including the title, x-axis and y-axis labels, figure width and height, horizontal gridlines, and x-axis label rotation. Display the final Plotly chart.

## import_libraries

### Subtask:
Import `pandas` for data manipulation and `plotly.express` for plotting.


**Reasoning**:
Import the `plotly.express` library as `px` for plotting, as specified in the instructions.



In [2]:
import plotly.express as px
print("plotly.express imported successfully as px.")

plotly.express imported successfully as px.


## define_chart_parameters

### Subtask:
Define the color map and prepare the legend labels including credit costs for each design based on the specifications.


**Reasoning**:
First, I will extract unique designs and their credit costs from the dataframe. Then, I will create a dictionary to map each design to its credit cost, formatted for legend labels, and define a custom color map for each design. These will be used for the upcoming visualization.



In [3]:
unique_designs_credits = df[['design', 'credits']].drop_duplicates().set_index('design')['credits'].to_dict()

design_credit_map = {
    design: f"{design} (credits: {credits})"
    for design, credits in unique_designs_credits.items()
}

color_map = {
    'design_a': '#1f77b4',
    'design_b': '#ff7f0e',
    'design_c': '#2ca02c'
}

print("Unique designs and their credit costs:", unique_designs_credits)
print("Design credit map for legend labels:", design_credit_map)
print("Custom color map:", color_map)

Unique designs and their credit costs: {'design_a': 820, 'design_b': 960, 'design_c': 900, 'design_d': 1000}
Design credit map for legend labels: {'design_a': 'design_a (credits: 820)', 'design_b': 'design_b (credits: 960)', 'design_c': 'design_c (credits: 900)', 'design_d': 'design_d (credits: 1000)'}
Custom color map: {'design_a': '#1f77b4', 'design_b': '#ff7f0e', 'design_c': '#2ca02c'}


**Reasoning**:
The previous `color_map` is incomplete as it does not include a color for 'design_d', which was identified in the `unique_designs_credits`. I need to update the `color_map` to ensure all unique designs have a corresponding color, as specified in the instructions.



In [4]:
color_map = {
    'design_a': '#1f77b4',
    'design_b': '#ff7f0e',
    'design_c': '#2ca02c',
    'design_d': '#d62728' # Adding a color for design_d
}

print("Updated custom color map:", color_map)

Updated custom color map: {'design_a': '#1f77b4', 'design_b': '#ff7f0e', 'design_c': '#2ca02c', 'design_d': '#d62728'}


**Reasoning**:
Now that the color map and legend labels are defined, I will generate the grouped bar chart using `plotly.express` as specified in the main task. I will use the `df` DataFrame, `workload` on the x-axis, `ipc` on the y-axis, color by `design`, and set `barmode='group'`. I will also apply the custom color map and legend labels, and then add chart customizations like title, axis labels, figure size, gridlines, and x-axis label rotation.



In [5]:
fig = px.bar(df,
             x='workload',
             y='ipc',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'ipc': 'IPC (Instructions Per Cycle)',
                 'design': 'Design'
             },
             title='IPC Comparison Across Designs and Workloads',
             category_orders={'design': list(design_credit_map.keys())} # Ensure consistent order
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig.data[i].name = design_credit_map[design]

# Customize layout
fig.update_layout(
    xaxis_title='Workload',
    yaxis_title='IPC',
    width=1000, # Set figure width
    height=600, # Set figure height
    yaxis=dict(
        gridcolor='lightgray', # Horizontal gridlines
        gridwidth=1
    ),
    xaxis=dict(
        tickangle=-45 # Rotate x-axis labels
    )
)

fig.show()

## Summary:

### Data Analysis Key Findings

*   The analysis successfully compared Instruction Per Cycle (IPC) values across four distinct designs (`design_a`, `design_b`, `design_c`, `design_d`) against various workloads.
*   A custom color map was applied to visually differentiate the designs in the chart.
*   Legend labels were enhanced to include the credit cost for each design, providing additional context (e.g., "design\_a (credits: 820)").
*   A grouped bar chart was generated, effectively visualizing IPC performance for each design under different workloads, allowing for direct comparison.

### Insights or Next Steps

*   Analyze the IPC performance relative to the credit costs for each design. This could help identify designs that offer a good balance of performance and cost efficiency, or those that underperform relative to their credit cost.
*   Further investigation could involve adding error bars to the IPC values if standard deviations are available, or exploring the impact of other design parameters or workload characteristics on IPC.


# IPC per Credit (Performance-Cost Efficiency)

# Task
Generate a grouped bar chart using `plotly.express` to compare IPC per credit across different designs and workloads, coloring the bars by `design` and setting `workload` on the x-axis and `ipc_per_credit` on the y-axis, with `barmode='group'`. Apply the previously defined `color_map` and `design_credit_map`. Customize the chart layout with an appropriate title, x-axis and y-axis labels, figure width of 1000 pixels, figure height of 600 pixels, horizontal gridlines, rotate x-axis labels by -45 degrees, and format the y-axis to display values in scientific notation with one decimal place. Display the resulting Plotly chart.

## import_libraries

### Subtask:
Ensure `plotly.express` is imported for creating the chart.


## create_grouped_bar_chart

### Subtask:
Generate a grouped bar chart using `plotly.express` with `ipc_per_credit` on the y-axis, `workload` on the x-axis, and bars colored and grouped by `design`.


**Reasoning**:
I will generate a grouped bar chart using `plotly.express` to visualize 'ipc_per_credit' across different designs and workloads, applying the predefined color map, updating legend labels, and customizing the chart's appearance as per the instructions.



In [6]:
fig_ipc_per_credit = px.bar(df,
             x='workload',
             y='ipc_per_credit',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'ipc_per_credit': 'IPC per Credit (Performance-Cost Efficiency)',
                 'design': 'Design'
             },
             title='IPC per Credit Comparison Across Designs and Workloads',
             category_orders={'design': list(design_credit_map.keys())} # Ensure consistent order
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_ipc_per_credit.data[i].name = design_credit_map[design]

# Customize layout for better readability
fig_ipc_per_credit.update_layout(
    xaxis_title='Workload',
    yaxis_title='IPC per Credit',
    width=1000, # Set figure width
    height=600, # Set figure height
    yaxis=dict(
        gridcolor='lightgray', # Horizontal gridlines
        gridwidth=1
    ),
    xaxis=dict(
        tickangle=-45 # Rotate x-axis labels
    )
)

fig_ipc_per_credit.show()

**Reasoning**:
I need to update the chart layout to format the y-axis values in scientific notation with one decimal place, as specified in the task.



In [7]:
fig_ipc_per_credit = px.bar(df,
             x='workload',
             y='ipc_per_credit',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'ipc_per_credit': 'IPC per Credit (Performance-Cost Efficiency)',
                 'design': 'Design'
             },
             title='IPC per Credit Comparison Across Designs and Workloads',
             category_orders={'design': list(design_credit_map.keys())} # Ensure consistent order
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_ipc_per_credit.data[i].name = design_credit_map[design]

# Customize layout for better readability
fig_ipc_per_credit.update_layout(
    xaxis_title='Workload',
    yaxis_title='IPC per Credit',
    width=1000, # Set figure width
    height=600, # Set figure height
    yaxis=dict(
        gridcolor='lightgray', # Horizontal gridlines
        gridwidth=1,
        tickformat='.1e' # Format y-axis to scientific notation with one decimal place
    ),
    xaxis=dict(
        tickangle=-45 # Rotate x-axis labels
    )
)

fig_ipc_per_credit.show()

## Summary:

### Data Analysis Key Findings
*   A grouped bar chart was successfully generated to compare IPC per credit across different designs and workloads.
*   The chart effectively uses `workload` on the x-axis, `ipc_per_credit` on the y-axis, and `design` to color and group the bars, employing `barmode='group'`.
*   The `color_map` and `design_credit_map` were correctly applied, with legend labels dynamically updated to reflect credit costs for each design.
*   The figure was customized with a width of 1000 pixels and a height of 600 pixels, featuring horizontal gridlines, and x-axis labels rotated by -45 degrees for improved readability.
*   The y-axis was formatted to display values in scientific notation with one decimal place (e.g., 1.2e-06), enhancing the precision and readability of performance-cost efficiency metrics.

### Insights or Next Steps
*   This comprehensive visualization provides a clear comparative view of the performance-cost efficiency (IPC per credit) of different designs under varying workloads, enabling identification of optimal designs for specific scenarios.
*   Further analysis could involve ranking designs based on their average IPC per credit across all workloads or investigating specific workload-design combinations that show unusually high or low efficiency.


#GRAPH 3: Branch Misprediction Rate Comparison

# Task
Create a grouped bar chart using `plotly.express.bar` to visualize `branch_mispredict_rate` on the y-axis, `workload` on the x-axis, and `design` for coloring and grouping the bars. Use the `color_map` and `design_credit_map` defined previously. Customize the chart with the title 'Branch Misprediction Rate Comparison Across Designs and Workloads', x-axis label 'Workload', y-axis label 'Branch Misprediction Rate (%)', set the figure width to 1200 pixels and height to 600 pixels, add horizontal gridlines, and rotate x-axis labels by -45 degrees. Display the resulting Plotly chart.

## import_libraries

### Subtask:
Ensure `plotly.express` is imported for creating the chart.


## create_and_customize_grouped_bar_chart

### Subtask:
Generate a grouped bar chart to compare branch misprediction rates across designs and workloads, then apply all specified layout customizations.


**Reasoning**:
I will generate a grouped bar chart using `plotly.express` to compare 'branch_mispredict_rate' across designs and workloads, applying the predefined color map, updating legend labels, and customizing the chart's appearance as specified in the instructions.



In [8]:
fig_branch_mispredict = px.bar(df,
             x='workload',
             y='branch_mispredict_rate',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'branch_mispredict_rate': 'Branch Misprediction Rate (%)',
                 'design': 'Design'
             },
             title='Branch Misprediction Rate Comparison Across Designs and Workloads',
             category_orders={'design': list(design_credit_map.keys())} # Ensure consistent order
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_branch_mispredict.data[i].name = design_credit_map[design]

# Customize layout for better readability
fig_branch_mispredict.update_layout(
    xaxis_title='Workload',
    yaxis_title='Branch Misprediction Rate (%)',
    width=1200, # Set figure width
    height=600, # Set figure height
    yaxis={
        "gridcolor": 'lightgray', # Horizontal gridlines
        "gridwidth": 1
    },
    xaxis={
        "tickangle": -45 # Rotate x-axis labels
    }
)

fig_branch_mispredict.show()

## Summary:

### Data Analysis Key Findings
*   A grouped bar chart was successfully generated to visualize the `branch_mispredict_rate` on the y-axis, `workload` on the x-axis, and `design` for coloring and grouping.
*   The chart was titled 'Branch Misprediction Rate Comparison Across Designs and Workloads', with x-axis labeled 'Workload' and y-axis labeled 'Branch Misprediction Rate (\%)'.
*   The legend labels were updated to include the credit costs associated with each design.
*   The chart's layout was customized with a width of 1200 pixels, a height of 600 pixels, horizontal gridlines on the y-axis, and x-axis labels rotated by -45 degrees for improved readability.

### Insights or Next Steps
*   The generated visualization provides a clear comparative view of branch misprediction rates across different designs and workloads, which can aid in identifying performance bottlenecks or optimal design choices for specific applications.
*   Further analysis could involve exploring the correlation between branch misprediction rates and overall system performance, or examining the impact of different design architectures on this metric under varying workloads.


# L1 Data Cache Miss Rate

# Task
Generate a grouped bar chart using `plotly.express.bar` to visualize the L1D miss rate as a percentage on the y-axis, `workload` on the x-axis, and `design` for coloring and grouping the bars. Calculate `l1d_miss_rate` by multiplying the existing column by 100. Use the previously defined `color_map` and `design_credit_map`. Customize the chart with an appropriate title, x-axis label 'Workload', y-axis label 'L1D Miss Rate (%)', set the figure width to 1200 pixels and height to 600 pixels, add horizontal gridlines, and rotate x-axis labels for readability. Display the resulting Plotly chart.

## import_libraries

### Subtask:
Ensure `plotly.express` is imported for creating the chart.


## prepare_data

### Subtask:
Calculate the L1D miss rate as a percentage.


**Reasoning**:
To calculate the L1D miss rate as a percentage, I need to multiply the `l1d_miss_rate` column by 100.



In [9]:
df['l1d_miss_rate_percent'] = df['l1d_miss_rate'] * 100
print("L1D miss rate percentage calculated and stored in 'l1d_miss_rate_percent' column.")
df[['design', 'workload', 'l1d_miss_rate', 'l1d_miss_rate_percent']].head()

L1D miss rate percentage calculated and stored in 'l1d_miss_rate_percent' column.


design       workload  l1d_miss_rate  l1d_miss_rate_percent
0  design_a      basicmath       0.041959                 4.1959
1  design_a      bitcounts       0.000016                 0.0016
2  design_a          qsort       0.092196                 9.2196
3  design_a    susan_edges       0.005025                 0.5025
4  design_a  susan_corners       0.008301                 0.8301

## create_and_customize_grouped_bar_chart

### Subtask:
Generate a grouped bar chart using `plotly.express.bar` to visualize the L1D miss rate as a percentage across different designs and workloads, then apply all specified layout customizations.


**Reasoning**:
I will generate a grouped bar chart using `plotly.express` to visualize the 'l1d_miss_rate_percent' across designs and workloads, applying the predefined color map, updating legend labels, and customizing the chart's appearance as specified in the instructions.



In [10]:
fig_l1d_miss_rate = px.bar(df,
             x='workload',
             y='l1d_miss_rate_percent',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'l1d_miss_rate_percent': 'L1D Miss Rate (%)',
                 'design': 'Design'
             },
             title='L1D Miss Rate Comparison Across Designs and Workloads',
             category_orders={'design': list(design_credit_map.keys())}
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_l1d_miss_rate.data[i].name = design_credit_map[design]

# Customize layout for better readability
fig_l1d_miss_rate.update_layout(
    xaxis_title='Workload',
    yaxis_title='L1D Miss Rate (%)',
    width=1200,
    height=600,
    yaxis={
        "gridcolor": 'lightgray',
        "gridwidth": 1
    },
    xaxis={
        "tickangle": -45
    }
)

fig_l1d_miss_rate.show()

## Summary:

### Data Analysis Key Findings
*   The L1D miss rate was successfully calculated as a percentage by multiplying the existing `l1d_miss_rate` column by 100 and stored in a new column named `l1d_miss_rate_percent`.
*   A grouped bar chart was generated to visualize the L1D miss rate percentage across different designs and workloads.
*   The chart clearly displays 'workload' on the x-axis and 'L1D Miss Rate (%)' on the y-axis, with 'design' used for coloring and grouping bars.
*   Customizations included a title "L1D Miss Rate Comparison Across Designs and Workloads," specific axis labels, dynamic legend labels incorporating credit costs, and layout adjustments (1200x600 pixels, horizontal gridlines, and -45 degree x-axis label rotation for readability).

### Insights or Next Steps
*   This visualization allows for a direct comparison of L1D cache performance (miss rates) among different hardware designs under various workloads, highlighting which designs perform better or worse in specific scenarios.
*   Further analysis could involve correlating the observed L1D miss rates with the credit costs of each design to evaluate the performance-to-cost efficiency, potentially leading to recommendations for optimal design choices based on specific performance requirements and budget constraints.


# Commit Utilization Comparison

# Task
Generate a grouped bar chart using `plotly.express.bar` to visualize `commit_utilization_pct` on the y-axis, `workload` on the x-axis, and `design` for coloring and grouping the bars. Use the existing `color_map` and `design_credit_map`. Customize the chart with the title 'Commit Stage Utilization by Workload and Design', x-axis label 'Benchmark', y-axis label 'Commit Utilization (%)', figure width of 1200 pixels and height of 600 pixels, add horizontal gridlines, and rotate x-axis labels for readability. Display the resulting Plotly chart.

## import_libraries

### Subtask:
Ensure `plotly.express` is imported for creating the chart.


## create_grouped_bar_chart

### Subtask:
Generate a grouped bar chart using `plotly.express.bar` with `commit_utilization_pct` on the y-axis, `workload` on the x-axis, and bars colored and grouped by `design`.


**Reasoning**:
I will generate a grouped bar chart using `plotly.express.bar` to visualize `commit_utilization_pct` across designs and workloads, applying the predefined color map, updating legend labels, and customizing the chart's appearance as specified in the instructions.



In [11]:
fig_commit_util = px.bar(df,
             x='workload',
             y='commit_utilization_pct',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Benchmark',
                 'commit_utilization_pct': 'Commit Utilization (%)',
                 'design': 'Design'
             },
             title='Commit Stage Utilization by Workload and Design',
             category_orders={'design': list(design_credit_map.keys())}
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_commit_util.data[i].name = design_credit_map[design]

fig_commit_util.show()

**Reasoning**:
The previous code block generated the grouped bar chart and updated the legend labels, but it did not include the specified layout customizations for figure width, height, horizontal gridlines, and x-axis label rotation. I need to add these customizations to complete the subtask as per the instructions.



In [12]:
fig_commit_util = px.bar(df,
             x='workload',
             y='commit_utilization_pct',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Benchmark',
                 'commit_utilization_pct': 'Commit Utilization (%)',
                 'design': 'Design'
             },
             title='Commit Stage Utilization by Workload and Design',
             category_orders={'design': list(design_credit_map.keys())}
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_commit_util.data[i].name = design_credit_map[design]

# Customize layout for better readability
fig_commit_util.update_layout(
    xaxis_title='Benchmark',
    yaxis_title='Commit Utilization (%)',
    width=1200, # Set figure width
    height=600, # Set figure height
    yaxis={
        "gridcolor": 'lightgray', # Horizontal gridlines
        "gridwidth": 1
    },
    xaxis={
        "tickangle": -45 # Rotate x-axis labels
    }
)

fig_commit_util.show()

## Summary:

### Data Analysis Key Findings
*   A grouped bar chart was successfully generated using `plotly.express.bar` to visualize `commit_utilization_pct` by `workload` (on the x-axis) and `design` (for coloring and grouping bars).
*   The chart incorporated `color_map` for consistent design coloring and `design_credit_map` to update legend labels with credit costs, enhancing data context.
*   Customizations included setting the title to 'Commit Stage Utilization by Workload and Design', labeling the x-axis as 'Benchmark' and the y-axis as 'Commit Utilization (\%)'.
*   The figure dimensions were set to 1200 pixels width and 600 pixels height, with horizontal gridlines added to the y-axis and x-axis labels rotated by -45 degrees for improved readability.
*   All specified visual enhancements were successfully implemented, resulting in a comprehensive and readable visualization.

### Insights or Next Steps
*   This chart effectively allows for a direct comparison of commit stage utilization across different workloads for each design, enabling identification of designs that perform more efficiently under specific benchmarks.
*   Further analysis could involve identifying benchmarks or designs with unusually high or low commit utilization to investigate potential bottlenecks or areas for optimization.


# Speculation Overhead

# Task
Generate a grouped bar chart using `plotly.express.bar` to compare `speculation_overhead_pct` across different designs and workloads, coloring the bars by `design` and setting `workload` on the x-axis and `speculation_overhead_pct` on the y-axis, with `barmode='group'`. Apply the previously defined `color_map` and `design_credit_map` to the chart.

## import_libraries

### Subtask:
Ensure `plotly.express` is imported for creating the chart.


## create_grouped_bar_chart

### Subtask:
Generate a grouped bar chart using `plotly.express.bar` with `speculation_overhead_pct` on the y-axis, `workload` on the x-axis, and bars colored and grouped by `design`.


**Reasoning**:
I will generate a grouped bar chart using `plotly.express.bar` to visualize `speculation_overhead_pct` across designs and workloads, applying the predefined color map, and setting initial labels and title as specified.



In [13]:
fig_speculation_overhead = px.bar(df,
             x='workload',
             y='speculation_overhead_pct',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'speculation_overhead_pct': 'Speculation Overhead (%)',
                 'design': 'Design'
             },
             title='Speculation Overhead by Workload and Design',
             category_orders={'design': list(design_credit_map.keys())}
            )

fig_speculation_overhead.show()

**Reasoning**:
The previous code generated the chart but did not update the legend labels to include credit costs. I need to iterate through the designs in `design_credit_map` and update the legend names in the figure to provide the additional context of credit costs, similar to previous charts in this notebook.



In [14]:
fig_speculation_overhead = px.bar(df,
             x='workload',
             y='speculation_overhead_pct',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Workload',
                 'speculation_overhead_pct': 'Speculation Overhead (%)',
                 'design': 'Design'
             },
             title='Speculation Overhead by Workload and Design',
             category_orders={'design': list(design_credit_map.keys())}
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_speculation_overhead.data[i].name = design_credit_map[design]

fig_speculation_overhead.show()

## customize_chart_layout

### Subtask:
Apply chart customizations including the title 'Speculation Overhead by Workload and Design', x-axis label 'Benchmark', y-axis label 'Speculation Overhead (%)', figure width of 1200 pixels and height of 600 pixels, horizontal gridlines, and x-axis label rotation for readability. Also, update the legend labels to include credit costs.


**Reasoning**:
The previous code block generated the grouped bar chart and updated the legend labels, but it did not include the specified layout customizations for figure width, height, horizontal gridlines, and x-axis label rotation. I need to add these customizations to complete the subtask as per the instructions.



In [15]:
fig_speculation_overhead = px.bar(df,
             x='workload',
             y='speculation_overhead_pct',
             color='design',
             barmode='group',
             color_discrete_map=color_map,
             labels={
                 'workload': 'Benchmark',
                 'speculation_overhead_pct': 'Speculation Overhead (%)',
                 'design': 'Design'
             },
             title='Speculation Overhead by Workload and Design',
             category_orders={'design': list(design_credit_map.keys())}
            )

# Update legend labels to include credit costs
for i, design in enumerate(design_credit_map.keys()):
    fig_speculation_overhead.data[i].name = design_credit_map[design]

# Customize layout for better readability
fig_speculation_overhead.update_layout(
    xaxis_title='Benchmark',
    yaxis_title='Speculation Overhead (%)',
    width=1200, # Set figure width
    height=600, # Set figure height
    yaxis={
        "gridcolor": 'lightgray', # Horizontal gridlines
        "gridwidth": 1
    },
    xaxis={
        "tickangle": -45 # Rotate x-axis labels
    }
)

fig_speculation_overhead.show()

## display_chart

### Subtask:
Display the final Plotly chart.


## Summary:

### Data Analysis Key Findings

*   A grouped bar chart was successfully generated to visualize `speculation_overhead_pct` across various `workload`s and `design`s.
*   The chart correctly mapped `workload` to the x-axis, `speculation_overhead_pct` to the y-axis, and used `design` for grouping and coloring of bars, applying a predefined `color_map`.
*   Legend labels were enhanced to include credit costs associated with each design, utilizing the `design_credit_map` for improved context.
*   The final chart incorporated comprehensive layout customizations:
    *   Title: 'Speculation Overhead by Workload and Design'.
    *   X-axis label: 'Benchmark'.
    *   Y-axis label: 'Speculation Overhead (%)'.
    *   Figure dimensions: 1200 pixels width and 600 pixels height.
    *   Visual aids: Horizontal gridlines were added to the y-axis, and x-axis labels were rotated by -45 degrees to improve readability.

### Insights or Next Steps

*   The visualization effectively presents the `speculation_overhead_pct` for different designs across various workloads, allowing for direct comparison of performance trade-offs, especially considering the credit costs embedded in the legend.
*   Further analysis could involve identifying which design offers the best balance between low speculation overhead and acceptable credit cost, potentially leading to recommendations for optimal design choices.
